### PR TITLE
feat(evals): rag evaluation scripts, bug fix for rails casing

### PR DESCRIPTION
**Pull request by jlopatec**
This includes some llama index scripts the team has been working on, added to scripts directory. Includes some changes so relevance works for both traces and normal Notebook Llama (this support was added, to relevance)

This also include bug fixes for rails

### CODE REVIEW

**Comment by jlopatec** [comment]
Bug in Rails was (capitalization based):
"Irregular" was railing to "regular" wrong rail

   # Convert the inputs to lowercase for case-insensitive matching
    string_lower = string.lower()
    positive_rail_lower = positive_rail.lower()
    negative_rail_lower = negative_rail.lower()

    positive_pos, negative_pos = string_lower.find(positive_rail_lower), string_lower.find(negative_rail_lower)

**Comment by jlopatec** [comment]
Added: output_map & trace_data
to run_relevance_eval so it could run with the Llama Scripts, so when not "trace_data", run_relevance_eval still works.

**Review by mikeldking — COMMENTED** [review]
@jlopatec ideally we would split this into two PRs as the scripts are un-related to the changes in some sense. I'm gonna get the team to take a look at the function signature - I don't want the `evals` lib to be span specific but rather have mechanisms to glue the metrics together - otherwise the metrics will not compose together well.

File: src/phoenix/experimental/evals/functions/binary.py [hunk 1]
```diff
@@ -2,10 +2,12 @@
 from typing import List, Optional, Set, Union
 
 import pandas as pd
+import tiktoken
```
Comment (mikeldking) [inline src/phoenix/experimental/evals/functions/binary.py:5]: i think if we need this it needs to be in at least our experimental deps

File: src/phoenix/experimental/evals/functions/binary.py [hunk 2]
```diff
@@ -85,7 +96,11 @@ def run_relevance_eval(
         either booleans or None in the case where the LLM output could not be
         parsed.
     """
-
+    if not output_map:
+        output_map = {
+            RAG_RELEVANCY_PROMPT_RAILS_MAP[0]: True,  # "relevant":True
+            RAG_RELEVANCY_PROMPT_RAILS_MAP[1]: False,  # "irrelevant":False
+        }
```
Comment (mikeldking) [inline src/phoenix/experimental/evals/functions/binary.py:103]: The rails are ordered dicts for a reason - this seems like a strange re-mapping.

File: src/phoenix/experimental/evals/functions/binary.py [hunk 3]
```diff
@@ -100,24 +115,30 @@ def run_relevance_eval(
...
+    if trace_data:
+        exploded_df[retrieved_document_text_column_name] = [
+            document_data["document.content"] if document_data is not None else None
+            for document_data in exploded_df[retrieved_documents_column_name]
+        ]
+    else:
+        exploded_df[retrieved_document_text_column_name] = [
+            document_data if document_data is not None else None
+            for document_data in exploded_df[retrieved_documents_column_name]
+        ]
```
Comment (mikeldking) [inline src/phoenix/experimental/evals/functions/binary.py:127]: this needs to be extracted out of this function so that it can be made composable. Doing this imperatively like so won't scale. I'd think we extract this to make it composable and could optionally detect if the documents are just chunks or document objects - no need to force the parameterization.

File: src/phoenix/experimental/evals/functions/binary.py [hunk 4]
```diff
@@ -100,24 +115,30 @@ def run_relevance_eval(
     ].map(list)
 
     exploded_df = filtered_df.explode(retrieved_documents_column_name, ignore_index=False)
-    exploded_df[retrieved_document_text_column_name] = [
-        document_data["document.content"] if document_data is not None else None
-        for document_data in exploded_df[retrieved_documents_column_name]
-    ]
+    if trace_data:
```
Comment (mikeldking) [inline src/phoenix/experimental/evals/functions/binary.py:118]: people will not know to set this flag - we should detect it or rely on composition so that the function is more pure.

File: src/phoenix/experimental/evals/functions/binary.py [hunk 5]
```diff
@@ -100,24 +115,30 @@ def run_relevance_eval(
...
+        exploded_df[retrieved_document_text_column_name] = [
+            document_data["document.content"] if document_data is not None else None
+            for document_data in exploded_df[retrieved_documents_column_name]
```
Comment (mikeldking) [inline src/phoenix/experimental/evals/functions/binary.py:121]: this is not pure - it's mutating the dataframe - this is gonna cause problems because the user might not be expecting that data to be modified and when the data is grouped back the data will lose score and document IDs.

File: src/phoenix/experimental/evals/functions/binary.py (hunk 4 above)
Comment (mikeldking) [reply src/phoenix/experimental/evals/functions/binary.py:118]: This code-path needs a test to ensure the df in is the same as the df out.

File: src/phoenix/experimental/evals/functions/binary.py [hunk 6]
```diff
@@ -188,17 +209,28 @@ def _extract_rail(string: str, positive_rail: str, negative_rail: str) -> Option
...
+        string = "Irregular"
+        Output: "irregular"
+
     """
 
-    positive_pos, negative_pos = string.find(positive_rail), string.find(negative_rail)
+    # Convert the inputs to lowercase for case-insensitive matching
+    string_lower = string.lower()
+    positive_rail_lower = positive_rail.lower()
+    negative_rail_lower = negative_rail.lower()
```
Comment (mikeldking) [inline src/phoenix/experimental/evals/functions/binary.py:221]: could be a separate PR - Let's move this out so it can be merged

File: src/phoenix/experimental/evals/functions/binary.py [hunk 7]
```diff
@@ -211,3 +243,60 @@ def _extract_rail(string: str, positive_rail: str, negative_rail: str) -> Option
     elif negative_pos != -1:
         return negative_rail
     return None
+
+
+def retrieval_concat_and_truncate(values, encoding_name="gpt-4", max_tokens=7100):
```
Comment (mikeldking) [inline src/phoenix/experimental/evals/functions/binary.py:248]: let's extract into a separate PR - it also feels like a pretty brute force solution that can affect the output - what if the context is at the end of the text? I think we should consider the reduce approach instead.

File: src/phoenix/experimental/evals/functions/binary.py [hunk 8]
```diff
@@ -211,3 +243,60 @@ def _extract_rail(string: str, positive_rail: str, negative_rail: str) -> Option
...
+
+
+def num_tokens_from_string(string: str, encoding_name: str) -> int:
+    """Returns the number of tokens in a text string."""
+    encoding = tiktoken.encoding_for_model(encoding_name)
+    num_tokens = len(encoding.encode(string))
+    return num_tokens
+
+
+def truncate_to_max_tokens(text: str, max_tokens=7100, encoding_name="gpt-4"):
```
Comment (mikeldking) [inline src/phoenix/experimental/evals/functions/binary.py:294]: Magic number - what is 7100?

File: src/phoenix/experimental/evals/functions/binary.py [hunk 9]
```diff
@@ -211,3 +243,60 @@ def _extract_rail(string: str, positive_rail: str, negative_rail: str) -> Option
...
+
+    """
+    concatenated_value = concatenate_values(values)
+    truncated_value = truncate_to_max_tokens(
+        concatenated_value, max_tokens=max_tokens, encoding_name=encoding_name
+    )
+    return truncated_value
+
+
+def concatenate_values(values):
```
Comment (mikeldking) [inline src/phoenix/experimental/evals/functions/binary.py:274]: add types

File: src/phoenix/experimental/evals/functions/binary.py [hunk 10]
```diff
@@ -211,3 +243,60 @@ def _extract_rail(string: str, positive_rail: str, negative_rail: str) -> Option
...
     return None
+
+
+def retrieval_concat_and_truncate(values, encoding_name="gpt-4", max_tokens=7100):
+    """This is designed to be used on a row of a Pandas Dataframe column.
+       value = ["chunk", "chunk", chunk"]
+       It concats the chunks in a list to a str, the column can be used for Q&A Eval.
+       df['retrieved_context'].apply(lambda x: retrieval_concat_and_truncate(x))
+       It makes sure the string can fit in a model / token <min> string size,
+       drops the tokens that don't fit.
```
Comment (mikeldking) [inline src/phoenix/experimental/evals/functions/binary.py:254]: dropping the tokens feels scary - should consider a different approach like splitting and accumulating.

**Comment by mikeldking** [comment]
@jlopatec I'm gonna have @fjcasti1 extract the relevant parts and merge the parts as they get approved.

- [ ] merge in the scripts
- [ ] account for context window size
- [ ] account for rails upper-case / lower-case

**Review by mikeldking — CHANGES_REQUESTED** [review]
Splitting into smaller PRs to incorporate evals work first, then the scripts will go last.

File: src/phoenix/experimental/evals/functions/binary.py (hunk 4 above)
Comment (jlopatec) [reply src/phoenix/experimental/evals/functions/binary.py:118]: I like the idea of detecting, good idea.

File: src/phoenix/experimental/evals/functions/binary.py (hunk 3 above)
Comment (jlopatec) [reply src/phoenix/experimental/evals/functions/binary.py:127]: In both cases here we are dealing with Chunks just at different hierarchies in the objects based on trace data or not. 

Maybe the suggestion below of detecting trace_data is the right thing to do, prior to this section.  Might be a way to slightly clean it up based on that.

File: src/phoenix/experimental/evals/functions/binary.py (hunk 5 above)
Comment (jlopatec) [reply src/phoenix/experimental/evals/functions/binary.py:121]: This approach wasn't really put in by me, just matched what was being done or is in the code now. 

If you look further down, its' grouped back up and inserted into a column in the original dataframe, so I don't think you lose anything. 

Might be cleaner ways of doing this. 

    collapsed_df = exploded_df.groupby(exploded_df.index, axis="index").agg(
        {
            llm_relevance_column_name: list,
        }
    )
    output_df = pd.DataFrame(index=dataframe.index)
    output_df[llm_relevance_column_name] = None
    output_df.loc[collapsed_df.index, llm_relevance_column_name] = collapsed_df[
        llm_relevance_column_name
    ]
    return output_df[llm_relevance_column_name].tolist()

File: src/phoenix/experimental/evals/functions/binary.py (hunk 3 above)
Comment (mikeldking) [reply src/phoenix/experimental/evals/functions/binary.py:127]: Yeah I like the idea of structure detection - cc @fjcasti1

**Comment by mikeldking** [comment]
Duplicate of #1520
